core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -257,11 +333,110 @@ public CompletableFuture<List<AcquiredRecords>> acquire(
+                    acquireNewBatchRecords(memberId, firstBatch.baseOffset(), lastBatch.lastOffset())));
+            }
 
-        return future;
+            log.trace("Overlap exists with in-flight records. Acquire the records if available for"
+                + " the share group: {}-{}", groupId, topicIdPartition);
+            List<AcquiredRecords> result = new ArrayList<>();
+            // The fetched records are already part of the in-flight records. The records might
+            // be available for re-delivery hence try acquiring same. The request batches could
+            // be an exact match, subset or span over multiple already fetched batches.
+            for (Map.Entry<Long, InFlightBatch> entry : subMap.entrySet()) {
+                InFlightBatch inFlightBatch = entry.getValue();
+                // Compute if the batch is a full match.
+                boolean fullMatch = checkForFullMatch(inFlightBatch, firstBatch.baseOffset(), lastBatch.lastOffset());
+
+                if (!fullMatch || inFlightBatch.offsetState != null) {
+                    log.trace("Subset or offset tracked batch record found for share partition,"
+                            + " batch: {} request offsets - first: {}, last: {} for the share"
+                            + " group: {}-{}", inFlightBatch, firstBatch.baseOffset(),
+                        lastBatch.lastOffset(), groupId, topicIdPartition);
+                    if (inFlightBatch.offsetState == null) {
+                        // Though the request is a subset of in-flight batch but the offset
+                        // tracking has not been initialized yet which means that we could only
+                        // acquire subset of offsets from the in-flight batch but only if the
+                        // complete batch is available yet. Hence, do a pre-check to avoid exploding
+                        // the in-flight offset tracking unnecessarily.
+                        if (inFlightBatch.batchState() != RecordState.AVAILABLE) {
+                            log.trace("The batch is not available to acquire in share group: {}-{}, skipping: {}"
+                                    + " skipping offset tracking for batch as well.", groupId,
+                                topicIdPartition, inFlightBatch);
+                            continue;
+                        }
+                        // The request batch is a subset or per offset state is managed hence update
+                        // the offsets state in the in-flight batch.
+                        inFlightBatch.maybeInitializeOffsetStateUpdate();
+                    }
+                    acquireSubsetBatchRecords(memberId, firstBatch.baseOffset(), lastBatch.lastOffset(), inFlightBatch, result);
+                    continue;
+                }
+
+                // The in-flight batch is a full match hence change the state of the complete.
```

Comment:
nit: "of the complete" is not grammatical. I think "of the complete batch" perhaps.

Reply:
Missed it, thanks.